Add SMBUS driver API, Intel PCH SMBUS driver, sample and tests

finikorg opened the pull request:
On Architecture WG meeting we agreed to have SMBus API in their own tree to simplify I2C and SMBus.

PR includes following components after request from @galak:
- SMBus subsystem API
- Intel PCH SMBus driver
- Qemu PCH driver tests
- Native Posix emulation tests
- SMBus sample with SMBus shell support

Can be verified with qemu_x86_64:
```
$ west build -b qemu_x86_64 zephyr/samples/drivers/smbus/ -p -t run
...
*** Booting Zephyr OS build zephyr-v3.3.0-1231-g00a4a871b5ba ***
Start SMBUS shell sample qemu_x86_64
uart:~$ smbus 
smbus - smbus commands
Subcommands:
  quick            :SMBus Quick command
                    Usage: quick <device> <addr>
  scan             :Scan SMBus peripheral devices command
                    Usage: scan <device>
  byte_read        :SMBus: byte read command
                    Usage: byte_read <device> <addr>
  byte_write       :SMBus: byte write command
                    Usage: byte_write <device> <addr> <value>
  byte_data_read   :SMBus: byte data read command
                    Usage: byte_data_read <device> <addr> <cmd>
  byte_data_write  :SMBus: byte data write command
                    Usage: byte_data_write <device> <addr> <cmd> <value>
  word_data_read   :SMBus: word data read command
                    Usage: word_data_read <device> <addr> <cmd>
  word_data_write  :SMBus: word data write command
                    Usage: word_data_write <device> <addr> <cmd> <value>
  block_write      :SMBus: Block Write command
                    Usage: block_write <device> <addr> <cmd> [<byte1>, ...]
  block_read       :SMBus: Block Read command
                    Usage: block_read <device> <addr> <cmd>
...
uart:~$ smbus scan smbus0 
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
00:             -- -- -- -- -- -- -- -- -- -- -- -- 
10: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
20: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
30: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
40: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
50: 50 51 52 53 54 55 56 57 -- -- -- -- -- -- -- -- 
60: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
70: -- -- -- -- -- -- -- --                         
8 devices found on smbus0
...
```

Testing can be done with following command:

```
zephyrproject$ twister -c -t smbus -v --all
...
INFO    - 1126968/1126992 rpl_crb                   tests/drivers/smbus/smbus_api/drivers.smbus.api.debug PASSED (build)
INFO    - 1126969/1126992 ehl_crb                   tests/drivers/smbus/smbus_api/drivers.smbus.api.debug PASSED (build)
INFO    - 1126970/1126992 qemu_x86_64               tests/drivers/smbus/smbus_api/drivers.smbus.api.no_smbalert PASSED (qemu 2.384s)
INFO    - 1126971/1126992 qemu_x86_64               tests/drivers/smbus/smbus_api/drivers.smbus.api.no_host_notify PASSED (qemu 2.458s)
INFO    - 1126972/1126992 rpl_crb                   tests/drivers/smbus/smbus_api/drivers.smbus.api.no_host_notify PASSED (build)
INFO    - 1126973/1126992 ehl_crb                   tests/drivers/smbus/smbus_api/drivers.smbus.api.no_host_notify PASSED (build)
INFO    - 1126974/1126992 ehl_crb                   tests/drivers/smbus/smbus_api/drivers.smbus.api.no_smbalert PASSED (build)
INFO    - 1126975/1126992 rpl_crb                   tests/drivers/smbus/smbus_api/drivers.smbus.api.no_smbalert PASSED (build)
INFO    - 1126976/1126992 qemu_x86_64               tests/drivers/smbus/smbus_api/drivers.smbus.api.debug PASSED (qemu 5.045s)
INFO    - 1126977/1126992 rpl_crb                   tests/drivers/smbus/smbus_api/drivers.smbus.api_stats PASSED (build)
INFO    - 1126978/1126992 qemu_x86_64               tests/drivers/smbus/smbus_api/drivers.smbus.api_stats PASSED (qemu 3.604s)
INFO    - 1126979/1126992 ehl_crb                   tests/drivers/smbus/smbus_api/drivers.smbus.api_stats PASSED (build)
INFO    - 1126980/1126992 rpl_crb                   tests/drivers/smbus/smbus_api/drivers.smbus.api.access.mmio PASSED (build)
INFO    - 1126981/1126992 ehl_crb                   tests/drivers/smbus/smbus_api/drivers.smbus.api.access.mmio PASSED (build)
INFO    - 1126982/1126992 qemu_x86_64               tests/drivers/smbus/smbus_api/drivers.smbus.api    PASSED (qemu 2.284s)
INFO    - 1126983/1126992 rpl_crb                   tests/drivers/smbus/smbus_api/drivers.smbus.api    PASSED (build)
INFO    - 1126984/1126992 ehl_crb                   tests/drivers/smbus/smbus_api/drivers.smbus.api    PASSED (build)
INFO    - 1126985/1126992 native_posix              tests/drivers/smbus/smbus_emul/drivers.smbus.emul.peripheral.interrupt PASSED (native 0.139s)
INFO    - 1126986/1126992 native_posix              tests/drivers/smbus/smbus_emul/drivers.smbus.emul  PASSED (native 0.090s)
INFO    - 1126987/1126992 ehl_crb                   samples/drivers/smbus/sample.drivers.smbus.shell   PASSED (build)
INFO    - 1126988/1126992 rpl_crb                   samples/drivers/smbus/sample.drivers.smbus.shell   PASSED (build)
INFO    - 1126989/1126992 qemu_x86_64               samples/drivers/smbus/sample.drivers.smbus.shell   PASSED (qemu 2.148s)
INFO    - 1126990/1126992 rpl_crb                   tests/drivers/smbus/smbus_api/drivers.smbus.api_user PASSED (build)
INFO    - 1126991/1126992 ehl_crb                   tests/drivers/smbus/smbus_api/drivers.smbus.api_user PASSED (build)
INFO    - 1126992/1126992 qemu_x86_64               tests/drivers/smbus/smbus_api/drivers.smbus.api_user PASSED (qemu 2.283s)

INFO    - 2242 test scenarios (1126992 test instances) selected, 1126967 configurations skipped (1126967 by static filter, 0 at runtime).
INFO    - 25 of 1126992 test configurations passed (100.00%), 0 failed, 0 errored, 1126967 skipped with 0 warnings in 273.92 seconds
INFO    - In total 101 test cases were executed, 4539671 skipped on 552 out of total 552 platforms (100.00%)
INFO    - 9 test configurations executed on platforms, 16 test configurations were only built.
...
```

Fixes #37101

finikorg committed:
board: elkhart_lake: Add smbus to the board's DTS

Add PCH SMBus to Elkhart Lake device tree.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
samples: Add smbus shell sample application

Add smbus sample for testing SMBus commands.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

stephanosio commented:
Rebased to fix CI failure

teburd commented:
I wonder if the route might be to treat an smbus controller as a super set or specialization of i2c instead, much like the i3c API is doing.

Please take a look at how that is being done in #46087

finikorg commented:
> I wonder if a better route might be to treat an smbus controller as a super set or specialization of i2c instead, much like the i3c API is doing.
> 
> Please take a look at how that is being done in #46087

I wanted to make things simple. I think we only need those smbus functions to be added.

finikorg committed:
smbus: dts: Add bindings for SMBus controllers

Adds bindings for SMBus controllers.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
smbus: Introduce SMBus subsystem driver API

Introduces SMBus driver API for SMBus controllers.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
smbus: shell: Add SMBus subsystem shell module

Add SMBus shell module for testing. This allows arbitrary boards with
SMBus driver supported exploring the SMBus communication with
peripheral devices.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
drivers: smbus: Add Intel ICH / PCH SMbus driver

Add driver supporting Intel I/O Controller Hub (ICH), later renamed
to Intel Platform Controller Hub (PCH).

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg commented:
> Architecture WG meeting:
> 
> * @henrikbrixandersen and @nashif argue that this should be its own API
> * @andyross mentions that the quirks in both smbus and I2c would likely make it impractical to share code
> * @finikorg agrees to move the API out of `i2c.h` into its own

I pushed updated version of the API.

galak reviewed (dismissed):
Don't we need at least on driver implementation and sample?

finikorg committed:
smbus: dts: Add Intel PCH SMBus driver bindings

Add bindings for Intel PCH SMBus driver.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
tests: drivers: Add tests for Intel PCH SMBus driver

Add test testing Intel PCH SMBus driver using Qemu x86_64 with
machine q35 SMBus controller emulation and default "eeprom"-like
peripheral.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg commented:
> Don't we need at least on driver implementation and sample?

Hi @galak, I wanted to get feedback for the API. I will add also driver implementation.

finikorg committed:
drivers: smbus: Add Intel PCH driver register header

Add header file for Intel PCH SMBus driver.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
smbus: dts: Add bindings for SMBus peripherals

Add bindings for SMBus peripheral devices on a bus.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
tests: smbus: Add SMBus unit tests

Add SMBus unit tests which emulate SMBus controller and connected test
peripheral device. This allows to test also rare SMBus protocols like
Process Call and Block Process Call.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
drivers: smbus: pch: Add Block Process Call operation

Add SMBus Block Process Call implementation for Intel PCH SMBus
driver.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
smbus: doc: Add Documentation for SMBus subsystem API

Add some basic documentation and API reference for SMBus.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
smbus: Update CODEOWNERS and MAINTAINERS files

Update CODEOWNERS and MAINTAINERS files.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
smbus: Add Host Notify and smbalert API

Add API for SMBus Host Notify and SMBALERT.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
drivers: smbus: Implement Host Notify and SMBALERT for PCH

Implement SMBus Host Notify and SMBALERT for Intel PCH driver.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
board: raptor_lake: Add smbus to the board's DTS

Add PCH SMBus driver to Raptor Lake device tree.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
smbus: Refactor callbacks API

The callbacks API was similar to gpio / espi callbacks API. Refactor
it to use more intuitive names for set / remove callbacks.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
smbus: Convert to use runtime BDF lookup

Convert PCH SMBus driver, tests and samples to use new dynamic BDF
lookup.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
board: qemu_x86_64: Add SMBus DTS configuration

Add SMBus specific configuration to qemu_x86_64 device tree since
this board does not have issues with dynamic IRQ allocation.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg commented:
> Don't we need at least on driver implementation and sample?

Added Intel PCH driver and sample, can be verified with qemu_x86_64

MaureenHelm commented on drivers/smbus/intel_pch_smbus.c:
```diff
@@ -53,6 +54,21 @@ struct pch_data {
 
 	struct k_mutex mutex;
 	struct k_sem completion_sync;
+	const struct device *dev;
+
+	/* smbalert callback list */
+	sys_slist_t smbalert_cbs;
+	/* smbalert work */
+	struct k_work smb_alert_work;
+
+	/* Host Notify callback list */
+	sys_slist_t host_notify_cbs;
+	/* Host Notify work */
+	struct k_work host_notify_work;
+	/* Host Notify peripheral device address */
+	uint8_t notify_addr;
+	/* Host Notify data received */
+	uint16_t notify_data;
```
Wrap in `ifdef CONFIG_SMBUS_INTEL_PCH_HOST_NOTIFY`?

finikorg replied:
Added wrap and tests

MaureenHelm reviewed (dismissed):
Nice work

finikorg committed:
drivers: smbus: pch: Wrap respective code with ifdefs

Wrap code for SMBALERT and HOST_NOTIFY with configuration ifdefs.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
tests: smbus: Test disabling configuration options

Add tests with some configuration option disabled:
CONFIG_SMBUS_INTEL_PCH_HOST_NOTIFY and
CONFIG_SMBUS_INTEL_PCH_SMBALERT.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
tests: smbus: Fix testing for disabled configuration

When some configuration options are disabled, test for -ENOSYS.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

teburd reviewed (approved):
really clean and thorough pr
